deku-derive/src/macros/deku_read.rs

```diff
@@ -352,7 +362,36 @@ fn emit_enum(input: &DekuData) -> Result<TokenStream, syn::Error> {
         });
     }
 
-    // println!("{}", tokens.to_string());
+    // create `deku_id` function for retrieving at runtime the id value
+    let deku_id_id_type = if id_type.is_some() || input.ctx.is_none() {
+        id_type.unwrap()
+    } else {
+        &ctx_types
+    };
+    let deku_id = if deku_ids.is_empty() {
+        quote! {
+            impl #ident {
+                fn deku_id(&self) -> #deku_id_id_type {
+                    match self {
+                        _ => panic!("deku id for variant not found")
+                    }
+                }
+            }
+        }
+    } else {
+        quote! {
+            impl #ident {
+                fn deku_id(&self) -> #deku_id_id_type {
+                    match self {
+                        #(#deku_ids),*,
```

Comment:
Since this is the only difference in the if case, I think we can collapse them into one
If `deku_ids` is empty, this should not even expand, leaving us with the first case :)

```suggestion
                        #(#deku_ids),*
```

Reply:
Well, with this diff:
```
diff --git a/deku-derive/src/macros/deku_read.rs b/deku-derive/src/macros/deku_read.rs
index b587736..54ffaf0 100644
--- a/deku-derive/src/macros/deku_read.rs
+++ b/deku-derive/src/macros/deku_read.rs
@@ -368,30 +368,19 @@ fn emit_enum(input: &DekuData) -> Result<TokenStream, syn::Error> {
     } else {
         &ctx_types
     };
-    let deku_id = if deku_ids.is_empty() {
-        quote! {
-            impl #ident {
-                fn deku_id(&self) -> #deku_id_id_type {
-                    match self {
-                        _ => panic!("deku id for variant not found")
-                    }
-                }
-            }
-        }
-    } else {
-        quote! {
-            impl #ident {
-                fn deku_id(&self) -> #deku_id_id_type {
-                    match self {
-                        #(#deku_ids),*,
-                        _ => panic!("deku id for variant not found"),
-                    }
+    let deku_id = quote! {
+        impl #ident {
+            fn deku_id(&self) -> #deku_id_id_type {
+                match self {
+                    #(#deku_ids),*
+                    _ => panic!("deku id for variant not found"),
                 }
             }
         }
     };
     tokens.extend(deku_id);
```
     
It looks like the last item doesn't print the trailing `,`
```
} impl TestEnumArray
{
    fn deku_id(& self) -> [u8 ; 3]
    {
        match self
        {
            Self :: VarA => * b"123", Self :: VarB => [1, 1, 1] _ => panic !
            ("deku id for variant not found"),
        }
    }
}
```

Reply:
Try this: `#(#var ,)*`

https://github.com/dtolnay/quote#repetition

Reply:
sweet, works. thanks.